### DISCUSSION TITLE
DevExtreme in Nuxt3 failing during build

### DISCUSSION THREAD

**Discussion by AloisSeckar** [nuxt/nuxt#16898] [Questions] [answered]
I am trying to include DevExtreme UI components into Nuxt 3 project. I took basic Nuxt3 app generated via `npx nuxi init nuxt-app` and started with bending it.

This is my test component with very basic DevExtreme menu:

    <template>
      <div id="container">
        <dxMenu>
            <dxItem icon="home" />
            <dxItem text="O aplikaci" />
            <dxItem icon="key" />
        </dxMenu>
      </div>
    </template>

    <script lang="ts">
      import DxMenu, { DxItem } from 'devextreme-vue/menu.js';
      import { defineComponent } from 'vue'
     
      export default defineComponent({
        components: {
            DxMenu, DxItem
        }
      })
    </script>

Then I just render the component in my `app.vue`. I have no other configuration related to that (apart from importing DevExtreme css file via `nuxt.config.ts`).

My `nuxt.config.ts` looks just like this:

    import { defineNuxtConfig } from 'nuxt'
    
    // https://v3.nuxtjs.org/api/configuration/nuxt.config
    export default defineNuxtConfig({
        css : [
            '@/assets/css/dx.generic.zdd-scheme.css'
        ]
    })

I was quite successful for dev environement - running `npm run dev` works like charm. But when I try `npm run build` it always fails.

The output looks like this:

    i Client built in 11430ms                                                13:38:43
    i Building server...                                                     13:38:43
    √ Server built in 4482ms                                                 13:38:47
    √ Generated public .output/public                                  nitro 13:38:47
    start Building server...                                           nitro 13:38:47
    
    ERROR  Rollup error: Error: Cannot find module devextreme/ui/menu imported from file:///C:/ATOS_Projekty/ZDD/web- 
    ui/app/node_modules, file:///C:/ATOS_Projekty/ZDD/web-ui/app/node_modules/nuxt/node_modules, 
    file:///C:/ATOS_Projekty/ZDD/web-ui/app/node_modules/nitropack/node_modules, file:///C:/ATOS_Projekty/ZDD/web-ui/app/, 
    file:///C:/ATOS_Projekty/ZDD/web-ui/app/node_modules/_index.js, file:///C:/ATOS_Projekty/ZDD/web-ui/app/node_modules, 
    file:///C:/ATOS_Projekty/ZDD/web-ui/app/node_modules/nuxt/, file:///C:/ATOS_Projekty/ZDD/web- 
    ui/app/node_modules/nuxt/node_modules/_index.js, file:///C:/ATOS_Projekty/ZDD/web-ui/app/node_modules/nuxt/node_modules, 
    file:///C:/ATOS_Projekty/ZDD/web-ui/app/node_modules/nitropack/, file:///C:/ATOS_Projekty/ZDD/web- 
    ui/app/node_modules/nitropack/node_modules/_index.js, file:///C:/ATOS_Projekty/ZDD/web- 
    ui/app/node_modules/nitropack/node_modules
    
    
    ERROR  Error: Cannot find module devextreme/ui/menu imported from file:///C:/ATOS_Projekty/ZDD/web-ui/app/node_modules, 
    file:///C:/ATOS_Projekty/ZDD/web-ui/app/node_modules/nuxt/node_modules, file:///C:/ATOS_Projekty/ZDD/web- 
    ui/app/node_modules/nitropack/node_modules, file:///C:/ATOS_Projekty/ZDD/web-ui/app/, file:///C:/ATOS_Projekty/ZDD/web- 
    ui/app/node_modules/_index.js, file:///C:/ATOS_Projekty/ZDD/web-ui/app/node_modules, file:///C:/ATOS_Projekty/ZDD/web- 
    ui/app/node_modules/nuxt/, file:///C:/ATOS_Projekty/ZDD/web-ui/app/node_modules/nuxt/node_modules/_index.js, 
    file:///C:/ATOS_Projekty/ZDD/web-ui/app/node_modules/nuxt/node_modules, file:///C:/ATOS_Projekty/ZDD/web- 
    ui/app/node_modules/nitropack/, file:///C:/ATOS_Projekty/ZDD/web-ui/app/node_modules/nitropack/node_modules/_index.js, 
    file:///C:/ATOS_Projekty/ZDD/web-ui/app/node_modules/nitropack/node_modules

    at resolvePath (/C:/ATOS_Projekty/ZDD/web-ui/app/node_modules/mlly/dist/index.mjs:1079:10)
    at _resolve (/C:/ATOS_Projekty/ZDD/web-ui/app/node_modules/nitropack/dist/shared/nitro.8d0213e3.mjs:304:22)
    at Object.resolveId (/C:/ATOS_Projekty/ZDD/web-ui/app/node_modules/nitropack/dist/shared/nitro.8d0213e3.mjs:348:34)
    at async resolveId (/C:/ATOS_Projekty/ZDD/web-ui/app/node_modules/rollup/dist/es/shared/rollup.js:21737:26)
    at async ModuleLoader.resolveId (/C:/ATOS_Projekty/ZDD/web-ui/app/node_modules/rollup/dist/es/shared/rollup.js:22037:19)
    at async /C:/ATOS_Projekty/ZDD/web-ui/app/node_modules/@rollup/plugin-commonjs/dist/es/index.js:773:14
    at async Promise.all (index 0)
    at async /C:/ATOS_Projekty/ZDD/web-ui/app/node_modules/@rollup/plugin-commonjs/dist/es/index.js:765:30
    at async rewriteRequireExpressionsAndGetImportBlock (/C:/ATOS_Projekty/ZDD/web-ui/app/node_modules/@rollup/plugin-commonjs/dist/es/index.js:1312:28)
    at async transformCommonjs (/C:/ATOS_Projekty/ZDD/web-ui/app/node_modules/@rollup/plugin-commonjs/dist/es/index.js:1874:23)`

`devextreme-vue/menu` is importing from `devextreme/ui/menu`, but it looks like this module is not being passed through during build. 

Do I have to declare the dependency somewhere in my code to notify build server to use it? I have tried various things already (like trying to convince Vite (rollup) not to perform tree shaking or to switch to Webpack instead), but always ending up with the same result. But maybe I just fail to do it correctly, I still have to learn a lot about how exactly the things work here. Any help/clues will be appreciated.

This is my `package.json` in case I shall change or install something there.

    {
      "private": true,
      "scripts": {
        "build": "nuxt build",
        "dev": "nuxt dev",
        "generate": "nuxt generate",
        "preview": "nuxt preview",
        "postinstall": "nuxt prepare",
        "start": "nuxt start"
      },
      "devDependencies": {
        "nuxt": "3.0.0-rc.8"
      },
      "dependencies": {
        "devextreme": "22.1.4",
        "devextreme-vue": "22.1.4"
      }
    }

EDIT: my **node** is `16.17.0` and **npm** `8.19.0`, maybe the versions are relevant...

**Comment by stenet** [comment, marked as answer]
add a file "dx.client.ts" to the ```plugins```-folder with the following example-content

```typescript
import { DxPopover } from "devextreme-vue/popover";
import { DxValidationGroup } from "devextreme-vue/validation-group";
import { DxDateBox, DxButton as DxDateBoxButton } from "devextreme-vue/date-box";
import { DxNumberBox, DxButton as DxNumberBoxButton } from "devextreme-vue/number-box";
import { DxValidationSummary } from "devextreme-vue/validation-summary";
import { DxTextArea } from "devextreme-vue/text-area";
import { DxPopup } from "devextreme-vue/popup";
import { DxButton } from "devextreme-vue/button";
import { DxTextBox, DxButton as DxTextBoxButton } from "devextreme-vue/text-box";
import { DxCheckBox } from "devextreme-vue/check-box";
import { DxValidator, DxCustomRule } from "devextreme-vue/validator";
import { DxScrollView } from "devextreme-vue/scroll-view";
import { DxList } from "devextreme-vue/list";

export default defineNuxtPlugin((nuxtApp) => {
  nuxtApp.vueApp
    .component("DxButton", DxButton)
    .component("DxTextBoxButton", DxTextBoxButton)
    .component("DxNumberBoxButton", DxNumberBoxButton)
    .component("DxDateBoxButton", DxDateBoxButton)
    .component("DxCheckBox", DxCheckBox)
    .component("DxNumberBox", DxNumberBox)
    .component("DxDateBox", DxDateBox)
    .component("DxTextBox", DxTextBox)
    .component("DxTextArea", DxTextArea)
    .component("DxValidator", DxValidator)
    .component("DxCustomRule", DxCustomRule)
    .component("DxList", DxList)
    .component("DxPopup", DxPopup)
    .component("DxPopover", DxPopover)
    .component("DxValidationGroup", DxValidationGroup)
    .component("DxValidationSummary", DxValidationSummary)
    .component("DxScrollView", DxScrollView);
});
```

I only got it working using ```ssr: false``` in the ```nuxt.config.ts```.

Now you can use these components in your vue components.

**Reply by stenet** [reply]
@AloisSeckar btw you have to use an uppercase D when using the components like

```html
<DxButton text="Hello" />
```

**Reply by AloisSeckar** [reply]
> add a file "dx.client.ts" to the plugins-folder with the following example-content

How do I reference the components in my .vue files then? import goes from where? 

Or should it "just work"? Because for me it doesn't... Getting:

    No value exists in scope for the shorthand property 'DxButton'. Either declare one or provide an initializer.

**Reply by stenet** [reply]
Yes, it just works ;)
See [https://stackblitz.com/github/stenet/nuxt-devextreme?file=app.vue](https://stackblitz.com/github/stenet/nuxt-devextreme?file=app.vue)

There you find a running sample (takes some time to start).

**Reply by stenet** [reply]
Just saw you have the problem when building. As far as I know this is still a known problem with DevExtreme. The only - for me - working solution is disabling tree-shaking => see the nuxt.config.ts in my sample.

**Reply by AloisSeckar** [reply]
ah, right...I was getting an error because I forgot to remove exporting DxButton as a component in `defineComponent` method of my `.vue` file...thanks a lot for guidance

**Reply by AloisSeckar** [reply]
> Just saw you have the problem when building. As far as I know this is still a known problem with DevExtreme. The only - for me - working solution is disabling tree-shaking => see the nuxt.config.ts in my sample.

It actually seems not to be necessary. After I created the plugin, it started working in build mode too. I don't have the `plugins` section of Vite configuration in my `nuxt.config.ts` and it works anyway.

**Reply by AloisSeckar** [reply]
Hi again, @stenet 
In case you'll have some time to help me again, I run into another problem related to that topic: https://github.com/nuxt/framework/discussions/7444

**Reply by AloisSeckar** [reply]
> I run into another problem related to that topic: nuxt/nuxt#16889

Never mind, it seems I figured it out. Turned out you were right in the end - disabling tree-shaking is actually required. It is not necessary for some DevExtreme UI components, but DxForm is one of those, where it has to be done.

The only possible question remaining - what is the exact reason for that? Is there some error in DevExtreme library? Or is it a Nuxt/Vite/Rollup issue?
